Add a base_url configuration setting to tts.

lddubeau opened the pull request:
## Description:

This adds a `base_url` configuration option for the `tts` component. For some usage scenarios, changing the `base_url` of the `http` component is not desirable. For instance, if someone is using a self-signed certificate to encrypt communications, the `say` service will not be able to play anything on a Google Home device, because the Google Home device will refuse to fetch from a `https://` url with a self-signed certificate. There's nothing that can be done on the Google Home end to fix this: you cannot install a new CA or do anything else. With the new `base_url` setting, it is possible to tell Google Home to fetch media at an `http://` address, and configure an `nginx` instance to serve only the URLs produced by `say` as `http` instead of `https`.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#6219

## Example entry for `configuration.yaml` (if applicable):

```yaml
tts:
  - platform: google
    base_url: http://192.168.0.10:8123
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)
  - [x] Tests have been added to verify that the new code works.

lddubeau committed:
Add a base_url configuration setting to tts.

homeassistant commented:
Hi @lddubeau, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#16478).

Once you do that we will be able to review and accept this pull request.

Thanks!

awarecan commented:
Thanks for the contribution. Please look the discussion in #15934

pvizeli commented:
How do you want trust your self generate certificate from outside? Do you print out the fingerprint of that certificate and take that on your body and check this fingerprint every time you access to this instance? Otherwise you don't need encrypt the Communication if you can't be trust it.

I know a certificate normally expensive but there also free alternative around like https://letsencrypt.org/ that allow you to create certificate they are trusted by a lot of device and help you to communicate secure and easy over the internet with your instance.

pvizeli commented:
Before we start a discussion -> https://github.com/home-assistant/home-assistant/pull/15934#issuecomment-419587314

lddubeau commented:
@pvizeli *By design*, my HA setup is not accessible from the outside. All the devices that are meant to access my HA server have the necessary information to verify that the certificate is correct. There's no need to print fingerprints. I've used Let's Encrypt since its early days.

pvizeli commented on homeassistant/components/tts/__init__.py:
```diff
@@ -180,13 +187,16 @@ def __init__(self, hass):
         self.use_cache = DEFAULT_CACHE
         self.cache_dir = DEFAULT_CACHE_DIR
         self.time_memory = DEFAULT_TIME_MEMORY
+        self.base_url = DEFAULT_BASE_URL
```
set it None

pvizeli commented:
@lddubeau that is the use case that I mean. So remove the SSL stuff from your comment and say what you realy need.

pvizeli commented:
@lddubeau fix the issue and we merge it

lddubeau commented:
I'll get on it soon.

lddubeau committed:
Remove the empty string as default value for base_urls

As requested in

https://github.com/home-assistant/home-assistant/pull/16478#pullrequestreview-153526144

lddubeau commented:
I've fixed the issues reported in the review.

eagappanath commented:
Hi lddubeau,
I use TTS for door status notification. After updating HA to 0.78.0, TTS has stopped working! Please see the error and current config below! After I read your post above, I did add base_url: http://192.168.1.5:8123 as well and it still doesn't work. Am I missing anything here? Could you please help me out? 

Thank you!

2018-09-19 12:26:49 ERROR (MainThread) [homeassistant.core] Error executing service <ServiceCall tts.google_say (c:5d80cef1b2224e4892d522d8aa6b4bbc): entity_id=['media_player.living_room_speaker'], message=Patio door closed.>
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/homeassistant/core.py", line 1127, in _event_to_service_call
    await service_handler.func(service_call)
  File "/usr/local/lib/python3.6/site-packages/homeassistant/components/tts/__init__.py", line 137, in async_say_handle
    options=options
  File "/usr/local/lib/python3.6/site-packages/homeassistant/components/tts/__init__.py", line 306, in async_get_url
    engine, key, message, use_cache, language, options)
  File "/usr/local/lib/python3.6/site-packages/homeassistant/components/tts/__init__.py", line 318, in async_get_tts_audio
    message, language, options)
  File "/usr/local/lib/python3.6/site-packages/homeassistant/components/tts/google.py", line 83, in async_get_tts_audio
    token.calculate_token, part)
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.6/site-packages/gtts_token/gtts_token.py", line 28, in calculate_token
    seed = self._get_token_key()
  File "/usr/local/lib/python3.6/site-packages/gtts_token/gtts_token.py", line 62, in _get_token_key
    a = re.search("a\\\\x3d(-?\d+);", tkk_expr).group(1)
AttributeError: 'NoneType' object has no attribute 'group'

tts:
  - platform: google
    cache: true
    cache_dir: /tmp/tts
    time_memory: 300

lddubeau commented:
@eagappanath The stack trace shows an error in the code that gets a token from https://translate.google.com My PR did not touch that code nor does it interact with that code.

MartinHjelmare commented:
Please open an issue if you suspect a bug. If you need help please use our help channels:
https://home-assistant.io/help/#communication-channels

Merged PRs should not be used for support or bug reports. Thanks!
